crates/noirc_evaluator/src/ssa/optim.rs

```diff
@@ -229,75 +155,82 @@ pub fn find_similar_instruction_with_multiple_arguments(
 
 pub fn find_similar_cast(
     igen: &SsaContext,
-    lhs: NodeId,
+    operator: &Operation,
     res_type: node::ObjectType,
     prev_ins: &VecDeque<NodeId>,
 ) -> Option<NodeId> {
     for iter in prev_ins {
         if let Some(ins) = igen.try_get_instruction(*iter) {
-            if ins.lhs == lhs && ins.res_type == res_type {
+            if &ins.operator == operator && ins.res_type == res_type {
                 return Some(*iter);
             }
         }
     }
     None
 }
+
 pub enum CseAction {
-    Replace, //replace the instruction
-    Remove,  //remove the instruction
-    Keep,    //keep the instruction
+    Replace { original: NodeId, replacement: NodeId },
+    Remove(NodeId),
+    Keep,
 }
 
-//Returns an id and an action:
-//- replace => the instruction should be replaced by the returned id
-//- remove  => the instruction corresponding to the returned id should be removed
-//- keep    => keep the instruction
-pub fn find_similar_mem_instruction(
+fn find_similar_mem_instruction(
     ctx: &SsaContext,
-    op: node::Operation,
+    op: &Operation,
     ins_id: NodeId,
-    lhs: NodeId,
-    rhs: NodeId,
-    anchor: &HashMap<node::Operation, VecDeque<NodeId>>,
-) -> (NodeId, CseAction) {
+    anchor: &mut Anchor,
```

Comment:
The function should take the queue of previous instructions, as find_similar_instruction does.

Reply:
That is what the Anchor is for

Reply:
no need to have it in this function, but this can be refactored later on.